numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -679,15 +679,62 @@ NPY_NO_EXPORT npy_bool
 PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
                                                     NPY_CASTING casting)
 {
-    /* Fast path for unsafe casts or basic types */
-    if (casting == NPY_UNSAFE_CASTING ||
-            (NPY_LIKELY(from->type_num < NPY_OBJECT) &&
+    /*
+     * Fast path for basic types.
+     */
+    if ((NPY_LIKELY(from->type_num < NPY_OBJECT) &&
              NPY_LIKELY(from->type_num == to->type_num) &&
              NPY_LIKELY(from->byteorder == to->byteorder))) {
         return 1;
     }
-    /* Equivalent types can be cast with any value of 'casting'  */
-    else if (PyArray_EquivTypenums(from->type_num, to->type_num)) {
+    /*
+     * Cases with fields need special treatment.
+     */
+    if (PyDataType_HASFIELDS(from)) {
+        /*
+         * If from is a structured data type, then it can be cast to a simple
+         * non-object one only for unsafe casting *and* if it has a single field.
+         */
+        if (!PyDataType_HASFIELDS(to) && to->type_num != NPY_OBJECT) {
+            return (casting == NPY_UNSAFE_CASTING &&
+                    PyDict_Size(from->fields) == 1);
+        }
+        /*
+         * Casting from one structured data type to another depends on the fields;
+         * we pass that case on to the EquivTypenums case below.
+         *
+         * TODO: move that part up here? Need to check whether equivalent type
+         * numbers is an addition constraint that is needed.
+         *
+         * TODO/FIXME: For now, always allow structured to structured for unsafe
+         * casting; this is not correct, but needed since the treatment in can_cast
+         * below got out of sync with astype; see gh-13667.
+         */
+        if (casting == NPY_UNSAFE_CASTING) {
+            return 1;
+        }
+    }
+    else if (PyDataType_HASFIELDS(to)) {
```

Comment:
What about object and subarrays? E.g `np.dtype("(1,2)u8")`

Reply:
Sorry, of course rather something like> `np.dtype(["f1", "f4", (5,)])` or `np.dtype([("f1", "f,f")])`

Reply:
I think for casting to structured array, sub-arrays do not matter - certainly- `astype` happily does it. This particular stanza is really mostly there to ensure we return `False` when casting is not "unsafe".

Reply:
For subarrays in `from`, `astype` seems quite happy to just take the first element. So, for now just mimicking that.

Reply:
What happens for:
```
from_dt = np.dtype([('a', ('i4,i4', 2))])
to_dt = np.dtype([('a', 'i4')])
```

For instance
```
arr = np.array(([(1, 2), (3, 4)],), dtype=dt)
arr.astype(to_dt)
```

What's currently implemented is not "just take the first element", but "take the first element and ignore all casting rules when casting it"

Reply:
Your particular example with `astype` raises an error in both in current master and in my PR, because `from_dt['a']` itself has two fields, which cannot be cast to a single one.

Sadly, though, `can_cast` happily claims this works even in my PR as `from['a']` has no fields (but does have a `subdtype`) - though only for `casting='unsafe'`.

Reply:
OK, that particular case is solved. But really `astype` is weird - it checks `can_cast` internally but can still throw an error later. This should not be allowed...

Reply:
Well, astype casting could fail based on the values. Of course you are right and it should not fail based on information already inside the dtype.

I have to catch up a bit here, sorry, but I do not really want crunch my mind too much about it right now, we are talking about corner cases and if we have new infrastructure things _might_ get easier to align.

Reply:
@seberg - agreed that we should take this bit-by-bit. My main goal really is to ensure that structured with multiple fields to unstructured always returns `False`, as this cannot possibly work (and has caused me real head-aches).